handler/handler.go

```diff
@@ -4,18 +4,63 @@ import (
 	"encoding/json"
 	"fmt"
 	"log"
+	"sync"
+	"time"
 
 	"github.com/nats-io/go-nats-streaming"
 	"github.com/openfaas/faas/gateway/queue"
 )
 
 // NatsQueue queue for work
 type NatsQueue struct {
-	nc        stan.Conn
-	ClientID  string
-	ClusterID string
-	NATSURL   string
-	Topic     string
+	nc                    stan.Conn
+	ncMutex               *sync.RWMutex
+	maxReconnect          int
+	delayBetweenReconnect time.Duration
+
+	ClientID              string
+	ClusterID             string
+	NATSURL               string
+	Topic                 string
+}
+
+func (q *NatsQueue) connect() error {
+	nc, err := stan.Connect(
+		q.ClusterID,
+		q.ClientID,
+		stan.NatsURL(q.NATSURL),
+		stan.SetConnectionLostHandler(func(conn stan.Conn, err error) {
+			log.Printf("Disconnected from %s\n", q.NATSURL)
+
+			q.reconnect(0)
+		}),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	q.ncMutex.Lock()
+	q.nc = nc
+	q.ncMutex.Unlock()
```

Comment:
@kozlovic do we want to use `defer q.ncMutex.Unlock()` instead to guarantee this happens?

Reply:
No, you don't want to include the publish in the lock, since it would prevent the reconnect logic to set the connection while the previous publish may be waiting for the timeout. The reason for the mutex is just to avoid a data race that would probably reported if you run `go test -race` with a test that would publish and have a disconnect event happening.

Reply:
Sorry, I thought this was in the Queue() API call. Regardless, no need for a defer here IMO.